clouddriver-kubernetes/src/main/java/com/netflix/spinnaker/clouddriver/kubernetes/config/BootstrapKubernetesConfigurationProvider.java

```diff
@@ -0,0 +1,206 @@
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import org.springframework.boot.context.properties.bind.BindResult;
+import org.springframework.boot.context.properties.bind.Bindable;
+import org.springframework.boot.context.properties.bind.Binder;
+import org.springframework.boot.context.properties.bind.PropertySourcesPlaceholdersResolver;
+import org.springframework.boot.context.properties.source.ConfigurationPropertySource;
+import org.springframework.boot.context.properties.source.MapConfigurationPropertySource;
+import org.springframework.cloud.bootstrap.config.BootstrapPropertySource;
+import org.springframework.context.ConfigurableApplicationContext;
+import org.springframework.core.env.ConfigurableEnvironment;
+import org.springframework.core.env.PropertySource;
+
+/**
+ * For larger number of Kubernetes accounts, as-is SpringBoot implementation of properties binding
+ * is inefficient, hence a custom logic for KubernetesConfigurationProperties is written but it
+ * still uses SpringBoot's Binder class. BootstrapKubernetesConfigurationProvider class fetches the
+ * flattened kubernetes properties from Spring Cloud Config's BootstrapPropertySource and creates a
+ * KubernetesConfigurationProperties object.
+ */
+public class BootstrapKubernetesConfigurationProvider {
+  private final ConfigurableApplicationContext applicationContext;
+  private CloudConfigResourceService configResourceService;
+  private SecretSession secretSession;
+  private Map<String, String> configServerCache;
+  private ObjectMapper objectMapper = new ObjectMapper();
+
+  public BootstrapKubernetesConfigurationProvider(
+      ConfigurableApplicationContext applicationContext,
+      CloudConfigResourceService configResourceService,
+      SecretManager secretManager) {
+    this.applicationContext = applicationContext;
+    this.configResourceService = configResourceService;
+    this.secretSession = new SecretSession(secretManager);
+  }
+
+  public KubernetesConfigurationProperties getKubernetesConfigurationProperties() {
+    return getKubernetesConfigurationProperties(getPropertiesMap());
+  }
+
+  @SuppressWarnings("unchecked")
+  public KubernetesConfigurationProperties getKubernetesConfigurationProperties(
+      Map<String, Object> kubernetesPropertiesMap) {
+    KubernetesConfigurationProperties k8sConfigProps = new KubernetesConfigurationProperties();
+    BindResult<?> result;
+
+    // unflatten
+    Map<String, Object> propertiesMap =
+        (Map<String, Object>)
+            JsonUnflattener.unflattenAsMap(kubernetesPropertiesMap).get("kubernetes");
+
+    // loop through each account and bind
+    for (Map<String, Object> unflattendAcc :
+        ((List<Map<String, Object>>) propertiesMap.get("accounts"))) {
+      result =
+          bind(getFlatMap(unflattendAcc), KubernetesConfigurationProperties.ManagedAccount.class);
```

Comment:
If I uderstand the code correctly, this is aproximately the sequence of transformations taking place:
```
Flattened Map<String, Object> --> Unflattened Map<String, Object> -->  (per account) String --> Flattened Map<String, Object> --> Binder --> Target class
```
Seems that the intermediate transformations are only used for handling `encrypted|encryptedFile|configserver` values by the `resolveSpecialCases` function.

In this case it's possible to use `ObjectMapper` to handle the special cases, and the transformation sequence could be simplied to:

```
Flattened Map<String, Object> --> Unflattened Map<String, Object> --> Target class
```

By only using `ObjectMapper` instead of `Binder` the code could look like:

```
Map<String, Object> unflattenedMap =
         (Map<String, Object>)
             JsonUnflattener.unflattenAsMap(kubernetesPropertiesMap).get("kubernetes");
KubernetesConfigurationProperties result = mapper.convertValue(unflattenedMap, KubernetesConfigurationProperties.class);
```

Of course individual account deserialization can be done with the same approach if that's a performance issue, but I haven't seen performance problems deserializing hundreds of accounts by `ObjectMapper` alone.

You can find an example of how to configure `ObjectMapper` to resolve special cases like `encrypted` and environment variables placeholders [here](https://github.com/armory-plugins/external-accounts/blob/master/clouddriver/src/main/java/io/armory/plugin/eap/loaders/URLCredentialsLoader.java#L81), the same approach can be done for `configserver`.

Reply:
>  you'll definitely need to re-implement parts of Spring to fix the issue, so definitely the solution will require that. I'm just trying to make sure this re-implements the minimal part 

@german-muzquiz I received the above review comment earlier hence I changed the code from using `ObjectMapper` to Spring Binder. The problems with SpingBoot binding surface when the whole `KubernetesConfigurationProperties` is supplied to SpringBoot instead of `ManagedAccount`s. So by using `Binder` class itself, we are achieving the performance just by changing the way we supply objects to it. This way I am able to keep the re-implementation to the minimum.

Reply:
I agree on the neccesity to reimplement parts of spring as part of this PR.
The issue I'm seeing is that you could have a big file having 1,000 k8s accounts, and you're doing several transformations of all the accounts that are not really needed, so instead of transforming from A to B you're transforming between intermediate types using the `getFlatMap` and `bind` methods.

`ObjectMapper` is able to transform between any given 2 types, so it very well can transform from `Map<String, Object>` of individual accounts to `ManagedAccount` like this:

```
Map<String, Object> unflattenedMap =
         (Map<String, Object>)
             JsonUnflattener.unflattenAsMap(kubernetesPropertiesMap).get("kubernetes");
List<Map<String, Object>> accountsMap = unflattenedMap.get("accounts");
List<ManagedAccount> acounts = accountsMap.stream()
  .map(a -> mapper.convertValue(a, ManagedAccount.class))
  .collect(Collectors::toList);
```

In this way we're still processing individual accounts instead of the full `KubernetesConfigurationProperties`, and we save the processing time of converting from unflatenned map to string and back to flatenned map that happens in `getFlatMap`.